### PR TITLE
feat(cli): scaffold infmonctl binary with full verb tree

### LINKED ISSUE #59: [W21] infmonctl CLI - full verb tree
## Spec
[007-cli](https://github.com/r12f/InFMon/blob/main/specs/007-cli.md)

## Scope
- Single Rust binary /usr/bin/infmonctl using clap
- Verb tree:
  - install/uninstall, start/stop/restart/status
  - config get/set/reload/show
  - flow-rule add/rm/list/show
  - flow list/show
  - stats show/export
  - log tail, health
- Connects to frontend via InFMonControlClient (shared ipc crate)
- Privilege model: root for writes, infmon group for reads via Unix socket

## Dependencies
Blocked by: #50

### PR DESCRIPTION
**Pull request by r12f**
## Summary

Implement the `infmonctl` CLI crate (`infmon-cli`) with clap-derived argument parsing for the complete v1 verb tree from [spec 007](https://github.com/r12f/InFMon/blob/main/specs/007-cli.md).

### What's included

- **Full verb tree**: install/uninstall, start/stop/restart/status, config get/set/reload/show, flow-rule add/rm/list/show, flow list/show, stats show/export, log tail, health
- **Global flags**: `--output table|json`, `--json`, `--compact`, `--raw-bytes`, `--config`, `--socket`, `--timeout`, `--no-color`, `-q`, `-v`
- **Exit codes module**: frozen v1 table (0, 1–7, 13, 69, 130, 143)
- **Privilege gates**: root-only verbs exit 13 for non-root callers
- **SIGPIPE handling**: clean exit 0 on broken pipe
- **Output helpers**: `TableDisplay` trait + JSON formatting utilities
- **38 integration tests**: argument parsing for every verb, flag combination, and rejection case

### What's NOT included (stubs)

All subcommand handlers are stubs returning exit 1. Actual frontend IPC integration (`InFMonControlClient`) will follow in subsequent issues (#60 builds on this).

Closes #59

### CODE REVIEW

File: src/cli/src/output.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    value: &T,
+    format: &OutputFormat,
+    compact: bool,
+) {
+    match format {
+        OutputFormat::Table => value.print_table(),
+        OutputFormat::Json => {
+            if compact {
+                println!("{}", serde_json::to_string(value).unwrap());
+            } else {
```
Comment (banidoru) [inline src/cli/src/output.rs:17]: `unwrap()` here will panic if serialization fails (e.g. map with non-string keys, or a `Serialize` impl that errors). Prefer `unwrap_or_else` with an error message to stderr + return, or at least document why it's safe for all expected types.

File: src/cli/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    },
+}
+
+#[derive(Subcommand, Debug)]
+pub enum FlowRuleCommands {
+    /// Add a flow rule (requires root)
+    Add {
+        /// Flow-rule spec in key=value form
+        spec: Vec<String>,
+    },
```
Comment (banidoru) [inline src/cli/src/lib.rs:159]: `spec: Vec<String>` allows `infmonctl flow-rule add` with zero specs, which would be a no-op. Consider `#[arg(required = true, num_args = 1..)]` to enforce at least one spec at the parse layer.

File: src/cli/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+// ---------------------------------------------------------------------------
+// Subcommand implementations (stubs that connect to the frontend via IPC)
+// ---------------------------------------------------------------------------
+
+async fn run_install(force: bool) -> i32 {
+    let _ = force;
+    // Check privilege
+    if !is_root() {
+        eprintln!("infmonctl: install requires root privileges");
+        return EXIT_PERMISSION_DENIED;
```
Comment (banidoru) [inline src/cli/src/main.rs:59]: All the `run_*` stub functions are `async fn` but perform no async work (no `.await`). This triggers `#[warn(unused)]`-family lints in some configurations. Consider making them plain `fn` returning `i32` until actual IPC is wired, or add a brief `// TODO: will await IPC client` comment to signal intent.

File: src/cli/src/output.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+use serde::Serialize;
+
+use crate::OutputFormat;
+
+/// Print a value as table or JSON based on the output format.
+pub fn print_output<T: Serialize + TableDisplay>(
```
Comment (banidoru) [inline src/cli/src/output.rs]: This function takes `compact` as a standalone param but doesn't accept `raw_bytes` or `no_color` — the caller will need to thread those through separately. Consider taking `&Cli` (or a small `OutputOptions` struct) so all formatting knobs are available in one place when the real rendering lands.

File: src/cli/tests/cli_args.rs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+            assert_eq!(spec, vec!["src_ip=10.0.0.0/8", "name=test"]);
+        }
+        _ => panic!("expected flow-rule add"),
+    }
+}
+
+#[test]
+fn parse_flow_rule_rm() {
+    let cli = parse(&["flow-rule", "rm", "test-rule"]);
+    match cli.command {
```
Comment (banidoru) [inline src/cli/tests/cli_args.rs:174]: Missing a test for `flow-rule add` with **zero** specs — this should either be rejected by clap or handled. Currently it silently parses as an empty vec.

**Review by banidoru — COMMENTED** [review]
- Overall solid scaffold — verb tree, exit codes, privilege gates, SIGPIPE handling all look correct
- `unwrap()` in `print_output` can panic on serialization failure; should handle gracefully
- `flow-rule add` accepts an empty spec vec — needs `#[arg(required = true)]` or clap `num_args(1..)`
- Stub functions are all `async` with no `.await` — consider making them sync until IPC lands to avoid dead-code warnings
- `output::print_output` signature takes `compact` but ignores `raw_bytes`/`no_color` from `Cli` — will be easy to forget wiring these later

File: src/cli/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+async fn run(cli: Cli) -> i32 {
+    let _output_format = cli.effective_output();
+
+    match cli.command {
+        Commands::Install { force } => run_install(force).await,
+        Commands::Uninstall { purge } => run_uninstall(purge).await,
+        Commands::Start => run_lifecycle("start").await,
+        Commands::Stop => run_lifecycle("stop").await,
+        Commands::Restart => run_lifecycle("restart").await,
```
Comment (banidoru) [inline src/cli/src/main.rs:39]: `_output_format` is computed here but never passed to any subcommand handler. Every `run_*` function independently ignores the format. This will become dead code unless the format is threaded through — consider passing it as a parameter now (or at least a `TODO` comment) so the plumbing is obvious when stubs get filled in.

File: src/cli/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    /// Per-call timeout when contacting the frontend (seconds)
+    #[arg(long, global = true, default_value_t = 5)]
+    pub timeout: u64,
+
+    /// Disable ANSI colors
+    #[arg(long, global = true, default_value_t = false)]
+    pub no_color: bool,
+
+    /// Suppress informational stderr
+    #[arg(short, long, global = true, default_value_t = false)]
```
Comment (banidoru) [inline src/cli/src/lib.rs:49]: `--timeout 0` is silently accepted but will almost certainly cause instant timeouts when IPC is wired up. Consider adding `#[arg(value_parser = clap::value_parser!(u64).range(1..))]` or documenting that 0 means "no timeout".

File: src/cli/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+// ---------------------------------------------------------------------------
+// Helpers
+// ---------------------------------------------------------------------------
+
+fn is_root() -> bool {
+    #[cfg(unix)]
+    {
+        unsafe { libc::geteuid() == 0 }
+    }
```
Comment (banidoru) [inline src/cli/src/main.rs:225]: On non-unix, `is_root()` unconditionally returns `false`, meaning all privilege-gated commands will always fail with exit 13. Worth a brief doc comment clarifying this is intentional (DPU-only target) so future contributors don't wonder if it's a bug.

**Review by banidoru — COMMENTED** [review]
- Scaffold looks solid — full verb tree, exit codes, privilege gates, SIGPIPE, 38 tests. Good foundation.
- `_output_format` computed but never threaded to handlers — dead code today, will need plumbing.
- `--timeout 0` is accepted but will likely cause instant timeouts; consider a minimum or validation.
- Minor: `is_root()` always returns `false` on non-unix — fine for DPU-only, but a doc comment would clarify intent.

File: src/cli/src/output.rs (hunk 1 above)
Comment (r12f) [reply src/cli/src/output.rs:17]: Good catch — replaced `unwrap()` with `unwrap_or_else` that prints a JSON error object to stdout instead of panicking. This keeps the output machine-parseable even on serialization failure. Fixed in 83233f0.

File: src/cli/src/lib.rs (hunk 2 above)
Comment (r12f) [reply src/cli/src/lib.rs:159]: Agreed — added `#[arg(required = true, num_args = 1..)]` to enforce at least one spec at the parse layer. Also added a `reject_flow_rule_add_without_spec` test to cover this. Fixed in 83233f0.

File: src/cli/src/main.rs (hunk 3 above)
Comment (r12f) [reply src/cli/src/main.rs:59]: Valid point. These are intentionally `async fn` because they will `.await` IPC client calls once wired — updated the section comment to clarify this intent: "stubs — will await IPC client when wired". Keeping them async avoids a churn commit when IPC lands.

File: src/cli/src/output.rs (hunk 4 above)
Comment (r12f) [reply src/cli/src/output.rs]: Reasonable suggestion for when real rendering lands. For now the only consumer is `print_output` which threads the knobs explicitly. I'll defer the `OutputOptions` struct to the IPC wiring PR where the real rendering logic will need all the flags in one place — avoids premature abstraction at the stub stage.

File: src/cli/tests/cli_args.rs (hunk 5 above)
Comment (r12f) [reply src/cli/tests/cli_args.rs:174]: Added `reject_flow_rule_add_without_spec` test and enforced `required = true` on the arg so clap rejects zero specs at parse time. Fixed in 83233f0.

File: src/cli/src/main.rs (hunk 6 above)
Comment (r12f) [reply src/cli/src/main.rs:39]: Acknowledged — `_output_format` is intentionally computed-but-unused at this stub stage to verify the flag-resolution logic works (tested in `parse_json_flag` and `parse_output_json`). Each `run_*` handler will receive it (or access it via `&Cli`) when IPC rendering is wired. The leading underscore signals this. No code change needed now.

File: src/cli/src/lib.rs (hunk 7 above)
Comment (r12f) [reply src/cli/src/lib.rs:49]: Good edge case. For this scaffold PR, `--timeout 0` is harmless since nothing does IPC yet. I'll add a `value_parser = clap::value_parser!(u64).range(1..)` constraint in the IPC wiring PR where it actually matters — adding it now would be untestable dead validation. Noting for follow-up.

File: src/cli/src/main.rs (hunk 8 above)
Comment (r12f) [reply src/cli/src/main.rs:225]: Correct — InFMon targets Linux/BlueField-3 exclusively (the spec doesn't cover Windows). The `#[cfg(not(unix))]` branch is a compile-time fallback to avoid build errors if someone opens the workspace on a non-unix host; it's not a supported runtime path. Adding a doc comment to clarify.

File: src/cli/src/output.rs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+/// Print a value as table or JSON based on the output format.
+pub fn print_output<T: Serialize + TableDisplay>(value: &T, format: &OutputFormat, compact: bool) {
+    match format {
+        OutputFormat::Table => value.print_table(),
+        OutputFormat::Json => {
+            if compact {
+                println!(
+                    "{}",
+                    serde_json::to_string(value)
+                        .unwrap_or_else(|e| format!("{{\"error\":\"serialize failed: {e}\"}}")),
```
Comment (banidoru) [inline src/cli/src/output.rs]: Nit: `format!("{{\"error\":\"serialize failed: {e}\"}}")` will produce invalid JSON if the serde error message contains `"` or `\`. Safer alternative:
```rust
.unwrap_or_else(|e| serde_json::json!({"error": format!("serialize failed: {e}")}).to_string())
```
Non-blocking — unlikely to hit in practice with current types.

**Review by banidoru — APPROVED** [review]
Iteration 2 re-review — all prior threads resolved. New commit cleanly addresses feedback:

- `unwrap()` → `unwrap_or_else` with error JSON fallback ✅
- `flow-rule add` now enforces `required = true, num_args = 1..` ✅
- New `reject_flow_rule_add_without_spec` test ✅
- rustfmt applied throughout ✅
- Async stub comment clarified ✅

One minor nit: the `unwrap_or_else` error fallback in `output.rs` uses `format!` to build JSON — if the serde error message ever contains quotes or backslashes, the output won't be valid JSON. Consider `serde_json::json!({"error": e.to_string()}).to_string()` for bulletproof output. Not blocking.

File: src/cli/src/output.rs (hunk 9 above)
Comment (banidoru) [reply src/cli/src/output.rs]: Still using `format!()` with manual brace escaping — the concern about invalid JSON when the serde error contains `"` or `\\` still applies. Non-blocking nit, but the `serde_json::json!({...}).to_string()` approach would be safer. Leaving open.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 2):

- Previous concerns addressed: `unwrap()` → `unwrap_or_else`, `spec` enforced `required = true`, intent comments added for async stubs and `_output_format`
- One prior nit still open: `format!()`-based error JSON in output.rs can produce invalid JSON on special chars — non-blocking
- New code is clean: CLI scaffold, exit codes, test coverage all solid
- No new issues found in this iteration

LGTM.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

File: src/cli/src/output.rs (hunk 9 above)
Comment (r12f) [reply src/cli/src/output.rs]: Fixed — switched both compact and pretty paths to use `serde_json::json!({"error": ...}).to_string()` so the fallback JSON is always valid regardless of what characters the serde error contains. Fixed in e22bd35.

**Review by banidoru — APPROVED** [review]
All prior review concerns addressed. LGTM.

- `unwrap()` → `unwrap_or_else` with `serde_json::json!` fallback — clean fix
- `flow-rule add` now enforces `required = true, num_args = 1..` with test coverage
- Async stub intent clarified in comments
- No new issues found in iteration 3 diff
